ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala

```diff
@@ -0,0 +1,483 @@
+    TemplatedStorageBackend.activeContractsEventsSameTemplates(
+      startExclusive = startExclusive,
+      endInclusiveSeq = endInclusiveSeq,
+      endInclusiveOffset = endInclusiveOffset,
+      templateIds = templateIds,
+      filteredWitnessesClause = arrayIntersectionValues("active_cs.flat_event_witnesses", parties),
+      submittersInPartiesClause = arrayIntersectionWhereClause("active_cs.submitters", parties),
+      witnessesWhereClause =
+        arrayIntersectionWhereClause("active_cs.flat_event_witnesses", parties),
+      limitExpr = limitClause(limit),
+      fetchSizeHint = fetchSizeHint,
+    )(connection)
+
+  def activeContractsEventsMixedTemplates(
+      startExclusive: Long,
+      endInclusiveSeq: Long,
+      endInclusiveOffset: Offset,
+      partiesAndTemplateIds: Set[(Ref.Party, Ref.Identifier)],
+      limit: Option[Int],
+      fetchSizeHint: Option[Int],
+  )(connection: Connection): Vector[EventsTable.Entry[Raw.FlatEvent]] = {
+    val parties = partiesAndTemplateIds.map(_._1)
+    TemplatedStorageBackend.activeContractsEventsMixedTemplates(
+      startExclusive = startExclusive,
+      endInclusiveSeq = endInclusiveSeq,
+      endInclusiveOffset = endInclusiveOffset,
+      filteredWitnessesClause = arrayIntersectionValues("active_cs.flat_event_witnesses", parties),
+      submittersInPartiesClause = arrayIntersectionWhereClause("active_cs.submitters", parties),
+      partiesAndTemplatesCondition = formatPartiesAndTemplatesWhereClause(
+        "active_cs.flat_event_witnesses",
+        partiesAndTemplateIds,
+      ),
+      limitExpr = limitClause(limit),
+      fetchSizeHint = fetchSizeHint,
+    )(connection)
+  }
+
+  def activeContractsEventsMixedTemplatesWithWildcardParties(
+      startExclusive: Long,
+      endInclusiveSeq: Long,
+      endInclusiveOffset: Offset,
+      partiesAndTemplateIds: Set[(Ref.Party, Ref.Identifier)],
+      wildcardParties: Set[Ref.Party],
+      limit: Option[Int],
+      fetchSizeHint: Option[Int],
+  )(connection: Connection): Vector[EventsTable.Entry[Raw.FlatEvent]] = {
+    val parties = wildcardParties ++ partiesAndTemplateIds.map(_._1)
+    TemplatedStorageBackend.activeContractsEventsMixedTemplatesWithWildcardParties(
+      startExclusive = startExclusive,
+      endInclusiveSeq = endInclusiveSeq,
+      endInclusiveOffset = endInclusiveOffset,
+      filteredWitnessesClause = arrayIntersectionValues("active_cs.flat_event_witnesses", parties),
+      submittersInPartiesClause = arrayIntersectionWhereClause("active_cs.submitters", parties),
+      witnessesWhereClause =
+        arrayIntersectionWhereClause("active_cs.flat_event_witnesses", wildcardParties),
+      partiesAndTemplatesCondition = formatPartiesAndTemplatesWhereClause(
+        "active_cs.flat_event_witnesses",
+        partiesAndTemplateIds,
+      ),
+      limitExpr = limitClause(limit),
+      fetchSizeHint = fetchSizeHint,
+    )(connection)
+  }
+
+  def flatTransactionSingleParty(
```

Comment:
What is the ultimate plan for methods like this one that is 100% identical between PG and H2. Wouldn't it be better to move it into the common implementation and let it be overridden only when necessary?

Reply:
Note that it's not 100% identical, because `arrayIntersectionWhereClause` is different for PG and H2.

Reply:
right, but that could be solve by making the method abstract in the common implementation

Reply:
yeah there is room for improvement here definitely, let's see how this looks after the first pass of oracle integration, and let's improve from there. template methds in a common-trait works, although is not my favorite, after we have all the code we can figure something nice I am sorue
also here is another dimension of improvment: squashing lots of event query methods to one, that would ultimately shrink the interface itself as well (as everywhere what implements it)